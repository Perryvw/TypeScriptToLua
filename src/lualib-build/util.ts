import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
