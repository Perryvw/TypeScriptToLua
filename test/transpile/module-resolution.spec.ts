import * as path from "path";
import * as util from "../util";

describe("basic module resolution", () => {
    const projectPath = path.resolve(__dirname, "module-resolution", "project-with-node-modules");

    const projectWithNodeModules = util
        .testProject(path.join(projectPath, "tsconfig.json"))
        .setMainFileName(path.join(projectPath, "main.ts"));

    test("can resolve global dependencies with declarations", () => {
        // Declarations in the node_modules directory
        expect(projectWithNodeModules.getLuaExecutionResult().globalWithDeclarationsResults).toEqual({
            foo: "foo from lua global with decls",
            bar: "bar from lua global with decls: global with declarations!",
            baz: "baz from lua global with decls",
        });
    });

    test("can resolve global dependencies with hand-written declarations", () => {
        // No declarations in the node_modules directory, but written by hand in project dir
        expect(projectWithNodeModules.getLuaExecutionResult().globalWithoutDeclarationsResults).toEqual({
            foo: "foo from lua global without decls",
            bar: "bar from lua global without decls: global without declarations!",
            baz: "baz from lua global without decls",
        });
    });

    test("can resolve module dependencies with declarations", () => {
        // Declarations in the node_modules directory
        expect(projectWithNodeModules.getLuaExecutionResult().moduleWithDeclarationsResults).toEqual({
            foo: "foo from lua module with decls",
            bar: "bar from lua module with decls: module with declarations!",
            baz: "baz from lua module with decls",
        });
    });

    test("can resolve module dependencies with hand-written declarations", () => {
        // Declarations in the node_modules directory
        expect(projectWithNodeModules.getLuaExecutionResult().moduleWithoutDeclarationsResults).toEqual({
            foo: "foo from lua module without decls",
            bar: "bar from lua module without decls: module without declarations!",
            baz: "baz from lua module without decls",
        });
    });

    test("can resolve package depencency with a dependency on another package", () => {
        // Declarations in the node_modules directory
        expect(projectWithNodeModules.getLuaExecutionResult().moduleWithDependencyResult).toEqual(
            "Calling dependency: foo from lua module with decls"
        );
    });

    test("resolved package dependency included in bundle", () => {
        const mainFile = path.join(projectPath, "main.ts");
        util.testProject(path.join(projectPath, "tsconfig.json"))
            .setMainFileName(mainFile)
            .setOptions({ luaBundle: "bundle.lua", luaBundleEntry: mainFile })
            .expectToEqual({
                globalWithDeclarationsResults: {
                    foo: "foo from lua global with decls",
                    bar: "bar from lua global with decls: global with declarations!",
                    baz: "baz from lua global with decls",
                },
                globalWithoutDeclarationsResults: {
                    foo: "foo from lua global without decls",
                    bar: "bar from lua global without decls: global without declarations!",
                    baz: "baz from lua global without decls",
                },
                moduleWithDeclarationsResults: {
                    foo: "foo from lua module with decls",
                    bar: "bar from lua module with decls: module with declarations!",
                    baz: "baz from lua module with decls",
                },
                moduleWithDependencyResult: "Calling dependency: foo from lua module with decls",
                moduleWithoutDeclarationsResults: {
                    foo: "foo from lua module without decls",
                    bar: "bar from lua module without decls: module without declarations!",
                    baz: "baz from lua module without decls",
                },
            });
    });
});

describe("module resolution with chained dependencies", () => {
    const projectPath = path.resolve(__dirname, "module-resolution", "project-with-dependency-chain");

    test("can resolve dependencies in chain", () => {
        //transpileProject(path.join(projectPath, "tsconfig.json"))
        util.testProject(path.join(projectPath, "tsconfig.json"))
            .setMainFileName(path.join(projectPath, "main.ts"))
            .expectToEqual({ result: "dependency3" });
    });

    test("resolved package dependency included in bundle", () => {
        const mainFile = path.join(projectPath, "main.ts");
        util.testProject(path.join(projectPath, "tsconfig.json"))
            .setMainFileName(mainFile)
            .setOptions({ luaBundle: "bundle.lua", luaBundleEntry: mainFile })
            .expectToEqual({ result: "dependency3" });
    });
});
