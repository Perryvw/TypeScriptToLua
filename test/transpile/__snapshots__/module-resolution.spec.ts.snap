// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports complicated paths configuration 1`] = `
Array [
  "/monorepo-complicated/dist/packages/tstl-program/packages/mypackage/src/bar.lua",
  "/monorepo-complicated/dist/packages/tstl-program/packages/mypackage/src/index.lua",
  "/monorepo-complicated/dist/packages/tstl-program/packages/myprogram/src/main.lua",
]
`;

exports[`supports paths configuration 1`] = `
Array [
  "/monorepo-with-paths/myprogram/dist/mypackage/bar.lua",
  "/monorepo-with-paths/myprogram/dist/mypackage/index.lua",
  "/monorepo-with-paths/myprogram/dist/myprogram/main.lua",
]
`;
